### ISSUE TITLE
[Linux/ARM64] 1-second stalls during the game with 60FPS patch

### ISSUE THREAD

**Issue by vanfanel** [sm64pc/sm64ex#422]
Hi there

I usually build sm64ex (nightly) with the 60FPS patch on my Raspberry Pi3b+.
I have a minimal system, no X server involved, and sm64ex runs on SDL2 using the very much lightweight SDL2 KMSDRM backend. There are no strange services interrumpting the CPU, only a minimal system based on Debian AARCH64.
SDL2 audio runs directly on ALSA (no PA or other bloatware).

The problem is that sm64ex is rock-solid 60FPS ALWAYS, but it shows 1-secong long stalls during gameplay. I have tried to determine the cause, without success. I have run the executable sm64.us.f3dex2e.arm from a RAMDISK to eliminate SD access from the equation, but these 1-second stalls still happen.

Maybe it's something related to audio buffering, so I tried setting a 1024 samples buffer in src/pc/audio/audio_sdl.c, but it does nothing for the stallls.

Any ideas? No one else noticed this?
Can be easily reproduced by leaving the demo running for a while (let it loop completely 1-2 times and it should happen).

**Comment by parkerlreed** [comment]
Link to the 60 fps patch? I can give this a try again on an AArch64 system.

**Comment by vanfanel** [comment]
@parkerlreed 
https://github.com/sm64pc/sm64ex/tree/nightly/enhancements

**Comment by vanfanel** [comment]
@parkerlreed I have been observing TOP, and when those 1-second stalls happen, there is NO process taking over the CPU (there was the possibility that some kernel thread with RT priority would be doing so): what happens is that **sm64 process simply uses 0% CPU for 1 second.** That leads me to think that there's some delay that's being triggered for an incorrect amount of time, or simply unneeded at all.

**Comment by fgsfdsfgs** [comment]
There are no explicit delays in the code other than the framerate control, which is turned off when vsync is enabled. It does not delay for more than ~33ms every frame, unless the implementation of `SDL_Delay` is screwed.
Have you tried forcing vsync with `--syncframes 1` or `--syncframes 2` (for 30 fps) or disabling it altogether in the config?
You can also try forcing the null audio driver by inserting a `return false;` at the very beginning of `audio_sdl_init` in `audio_sdl.c`.
As far as I can tell, this seems to only happen in your case, so on a RPi, which I don't have and don't know the hardware quirks of, so it could really be anything from weird vsync wait implementation to audio-related delays.

**Comment by vanfanel** [comment]
@parkerlreed I have completely disabled the only SDL_Delay() call I found (wasn't used anyway with vsync ON). Also, the code for vsync in SDL2 in the KMSDRM backend is mine, it's solid:
https://github.com/vanfanel/SDL/blob/58bbfacee72aa798d23ae3e6582db52d0676ba31/src/video/kmsdrm/SDL_kmsdrmopengles.c#L110

So I would say it's something related to audio, since it's the only game I have built for the Pi that outputs 32KHz audio. Is there a way to change it on the code?
I have looked in src/pc/audio/audio_sdl2.c: I changed buffer size, etc. BUt changing the audio rate to 44100Hz totally breaks the game audio as expected.

**Comment by fgsfdsfgs** [comment]
The mixer and some other things are tailored for 32khz, so not easily. Have you tried disabling the sound output altogether like I suggested?

**Comment by vanfanel** [comment]
@fgsfdsfgs Nope, next thing I will try. Could it be that the Pi is the only platform where SDL2 intefaces with ALSA directly instead of having PA under SDL2?

**Comment by fgsfdsfgs** [comment]
I don't know that, but the [original sm64-port](https://github.com/sm64-port/sm64-port) has raw ALSA and Pulse backends. Maybe you could try those and see if you get the same problem?

**Comment by vanfanel** [comment]
@fgsfdsfgs The problem is that the original sm64-port lacks OpenGL_ES support, instead relying on the rotten GLX implementation in the Linux platform, so there's no apparent way to build it for anything like SDL2/KMSDRM.

**Comment by parkerlreed** [comment]
Yeah I use pulseaudio everywhere. Haven't noticed any weird issues.
